Add strict kwarg to `nn.Module.set_submodule` and fix bug for non dot delineated strings

mariovas3 opened the pull request:
Before fixing set_submodule, it used to create leaf modules when the target was not a dot-delimited string. After the fix it will not create a new attribute if target is a non-dot-delimited string. If you want to create leaf nodes of `nn.Module` parent nodes, you can use `replace_or_create_new_leaf_module`.

Fixes https://github.com/pytorch/pytorch/issues/143441

mariovas3 committed:
fix set_submodule and add replace_or_create_new_leaf_module

before fixing set_submodule, it used to create leaf modules when the target was not a dot-delimited string. after the fix it will not create a new attribute if target is a non-dot-delimited string. If you want to create leaf nodes of nn.Module parent nodes, you can use replace_or_create_new_leaf_module

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -770,22 +770,83 @@ def set_submodule(self, target: str, module: "Module") -> None:
-                    mod._get_name() + " has no attribute `" + item + "`"
+        setattr(parent, atoms[-1], module)
+
+    def replace_or_create_new_leaf_module(self, target: str, module: "Module"):
```
Why do we add a new API here instead of updating the documentation of `set_submodule` to reflect its actual behavior (of setting the submodule regardless of whether it exists)?

mariovas3 replied:
Currently, `set_submodule` sets a module to a potentially non-existent attribute name only if the `target` is a non-dot-delimited string. This violates the docstring which says `Set the submodule given by target if it exists, otherwise throw an error.`

My understanding is that the docstring was the feature you guys wanted to support and the current (inconsistent) behaviour (depending on whether the `target` string is dot-delimited) is the bug.

You seem to be suggesting that the bug is a feature.

>Why do we add a new API

I saw that the undocumented behaviour (some may call it a bug) is used in the repository, suggesting the need for a method that adds a leaf node if it doesn't exist.

mariovas3 replied:
Admittedly, the current docstring (released on the 2.5 stable release) describes the feature I was personally looking for, so I tested it out when I saw it. That's why I thought the behaviour described above is a bug.

I agree, this is a slightly strange PR in that it is named bugfix - and yet also introduces a new API/feature. As mentioned above, I thought this was necessary due to the use of `set_submodule` in this repository (see changed files in this PR) in a way that is inconsistent with its docstring (as per 2.5 stable release https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.set_submodule).

Please advise on steps forward.

mikaylagawarecki replied:
Got it, thank you! I missed the part where it's only possible to set a nonexistent submodule if the `target` string is not dot-delimited

> the undocumented behaviour (some may call it a bug) is used in the repository,

Are the callsites above actually using the undocumented behavior? It appears to me that they are all attempting to set submodules that exist, but do correct me if I missed something.

If so I would suggest just correcting `set_submodule` to properly throw an error in that one bad case

mariovas3 replied:
> If so I would suggest just correcting set_submodule to properly throw an error in that one bad (non dot delimited) case and not adding a new replace_or_create_new_leaf_module API

That sounds good. From memory I tested using the recommended way e.g.,

```bash
python test/export/test_export.py TestExport.test_preserve_module_call_signature_unflatten_specialization
```
this line will break after the bugfix of `set_submodule`

https://github.com/pytorch/pytorch/pull/143455/files#diff-4a060a24e1f81389eab7390d434dddf919af50aa8fda6cbd81e182a53bd9328eL8473

I shall retest and let you know. If we indeed don't encounter problems with the `test_export.py`, it makes sense to remove the new api and confine the PR to the bugfix of `set_submodule`.

mariovas3 replied:
I reverted the changes to `torch/export/unflatten.py` and `test/export/test_export.py` (reverted to using the now fixed `set_submodule` instead of the new `replace_or_create_new_leaf_module` api)

result of 

```bash
python ../test/export/test_export.py TestExport.test_unflatten_no_unroll
```

```
ERROR: test_unflatten_no_unroll (__main__.TestExport.test_unflatten_no_unroll)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/focal/coding/pytorch/torch/testing/_internal/common_utils.py", line 3108, in wrapper
    method(*args, **kwargs)
  File "/home/focal/coding/pytorch/torch/../test/export/test_export.py", line 8430, in test_unflatten_no_unroll
    test(
  File "/home/focal/coding/pytorch/torch/../test/export/test_export.py", line 8416, in test
    ufm = torch.export.unflatten(ep)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/focal/coding/pytorch/torch/export/unflatten.py", line 695, in unflatten
    return UnflattenedModule(module, flat_args_adapter)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/focal/coding/pytorch/torch/export/unflatten.py", line 480, in __init__
    self._dispatch_modules(redirected_call_indices, consts_targets)
  File "/home/focal/coding/pytorch/torch/export/unflatten.py", line 632, in _dispatch_modules
    self.set_submodule(
  File "/home/focal/coding/pytorch/torch/nn/modules/module.py", line 781, in set_submodule
    raise AttributeError(
AttributeError: UnflattenedModule has no attribute `n`

To execute this test, run the following from the base repo dir:
    python test/export/test_export.py TestExport.test_unflatten_no_unroll

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 1 test in 1.018s

FAILED (errors=1)
```

mariovas3 replied:
ok, so I reverted the changes to the tests in `test/export/test_export.py` to use `set_submodule` again, but it does seem necessary to use `replace_or_create_new_leaf_module` in `torch/export/unflatten.py` in place of `set_submodule` in order for the tests in `test/export/test_export.py` to pass.

I tested with:

```
python ../test/export/test_export.py TestExport.test_unflatten_no_unroll
python ../test/export/test_export.py TestExport.test_preserve_module_call_signature_unflatten_specialization
python ../test/export/test_export.py TestExport.test_unflatten_multiple_graphs_preserve_signature_no_error
python ../test/export/test_export.py TestExport.test_unflatten_multiple_graphs_state
```

and everything runs without errors.

mikaylagawarecki replied:
Apologies for the delay, I would propose the following

- let's not add the new public API `replace_or_create_new_leaf_module`
- instead, `set_submodule` can take in a kwarg `strict` (similar to `strict` in `load_state_dict`)

if `strict=False`, do what `replace_or_create_new_leaf_module` does --
Replace the submodule given by ``target`` if it exists, or see if the parent module of target exists and create target, otherwise throw an error.

If `strict=True` use the changes to `set_submodule` with your bugfix -- 
Set the submodule given by ``target`` if it exists, otherwise throw an error

We can pass this kwarg as appropriate in the export logic

mariovas3 replied:
Happy New Year!

I like the solution you're proposing. I'll work on it.

mariovas3 replied:
Done.

* I added the `strict` kwarg and removed the `replace_or_create_new_leaf_module` api.
    * The unit tests for `test_export.py` now all pass when using `set_submodule` with the default `strict=False`.
* I wrote tests for `set_submodule` that are based on its docstring.

mariovas3 committed:
the tests for export/test_export.py don't need modifications as long as the export/unflatten.py module replaces torch.nn.Module.set_submodule with torch.nn.Module.replace_or_create_new_module

mariovas3 committed:
improve docs about when set_submodule and get_submodule raise an AttributeError

mariovas3 committed:
add strict=False kwarg to set_submodule and remove replace_or_create_new_leaf_module api

mariovas3 commented:
Hi @mikaylagawarecki , just a kind reminder that I have made the changes you asked for. Can you please provide feedback. Thanks.

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -740,52 +747,68 @@ def set_submodule(self, target: str, module: "Module") -> None:
+        To override the ``Conv2d`` with a new submodule ``Linear``, you
         would call
-        ``set_submodule("net_b.net_c.conv", nn.Linear(33, 16))``.
+        ``set_submodule("net_b.net_c.conv", nn.Linear(1, 1))`` or
+        ``set_submodule("net_b.net_c.conv", nn.Linear(1, 1), strict=False)`` or
+        ``set_submodule("net_b.net_c.conv", nn.Linear(1, 1), strict=True)``.
```
```suggestion
        you could call ``set_submodule("net_b.net_c.conv", nn.Linear(1, 1))`` where ``strict`` could be ``True`` or ``False``

mikaylagawarecki reviewed (commented):
Thanks a lot and thanks for your patience! Looks good if CI is green, I just have 2 doc nits

mariovas3 committed:
follow torch convention for docs

mariovas3 commented:
I fixed the docs, but that cancelled the CI :cry: 

Can you please approve the CI.

mikaylagawarecki commented:
@pytorchbot -r

mikaylagawarecki commented:
@pytorchbot rebase

mariovas3 commented:
Hi, I was about to mention I got some strange `GitHub Actions has encountered an internal error when running your job.` message on some of the workflows from the CI you kicked off yesterday @mikaylagawarecki . Not sure how to debug this.

mariovas3 commented:
I don't think the failing workflow broke because of my change. It looks like a classic output matching error from stochastic input.

As can be seen in the printout below, the location metrics - min, median and max - are the same but the mean is different. Unlikely an error caused by setting a module.

Another nit, the failing test violates the recommendation to use `make_tensor` instead of other tensor-creation APIs like `torch.randn`.

https://github.com/pytorch/executorch/blob/main/backends/xnnpack/test/ops/test_conv1d.py#L149
https://github.com/pytorch/pytorch/wiki/Running-and-writing-tests#common-test-utilities

```
=================================== FAILURES ===================================
___________________ TestConv1d.test_qs8_conv1d_batchnorm_seq ___________________
[gw0] linux -- Python 3.10.16 /opt/conda/envs/py_3.10/bin/python
Traceback (most recent call last):
  File "/executorch/backends/xnnpack/test/ops/test_conv1d.py", line 151, in test_qs8_conv1d_batchnorm_seq
    self._test_conv1d(
  File "/executorch/backends/xnnpack/test/ops/test_conv1d.py", line 117, in _test_conv1d
    tester.to_executorch().serialize().run_method_and_compare_outputs()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/executorch/backends/xnnpack/test/tester/tester.py", line 660, in run_method_and_compare_outputs
    self._compare_outputs(
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/executorch/backends/xnnpack/test/tester/tester.py", line 725, in _compare_outputs
    Tester._assert_outputs_equal(
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/executorch/backends/xnnpack/test/tester/tester.py", line 682, in _assert_outputs_equal
    assert torch.allclose(
AssertionError: Output 0 does not match reference output.
	Given atol: 0.001, rtol: 0.001.
	Output tensor shape: torch.Size([6, 2, 4]), dtype: torch.float32
	Difference: max: 0.011491775512695312, abs: 0.011491775512695312, mean abs error: 0.00047882398939691484.
	-- Model vs. Reference --
		 Numel: 48, 48
	Median: -0.7584591507911682, -0.7584591507911682
	  Mean: -0.6198393106460571, -0.6203181147575378
	   Max: 0.7929345369338989, 0.7929345369338989
	   Min: -1.9536068439483643, -1.9536068439483643
```

mariovas3 commented:
Hi @mikaylagawarecki , can you please advise on next steps.

mariovas3 committed:
change type hint from List[str] to list[str] in line with a pr called [5/N] Apply Ruff fixes

this came after I fixed a merge conflict (introduced by the aforementioned ruff pr) by accepting my incoming changes without changing List[str] to list[str]

mariovas3 commented:
Looks like everyone is failing this test - 
![bad_requirements_install](https://github.com/user-attachments/assets/16607ca6-4d13-4717-ab33-b635f54f3b30)

mikaylagawarecki commented:
@pytorchbot merge

mariovas3 commented:
Thanks for reviewing @mikaylagawarecki !

For completeness, this is the error of the only failing job at the time of merging:

job link - https://github.com/pytorch/pytorch/actions/runs/12785695690/job/35667949111?pr=143455

```
+ bash examples/models/llama3_2_vision/install_requirements.sh
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/nightly/cpu
ERROR: Could not find a version that satisfies the requirement torchtune==0.4.0.dev20241112 (from versions: 0.0.1, 0.1.0, 0.1.1, 0.2.0, 0.2.1, 0.3.0, 0.3.1, 0.4.0.dev20241113+cpu, 0.4.0.dev20241114+cpu, 0.4.0, 0.5.0.dev20241115+cpu, 0.5.0.dev20241116+cpu, 0.5.0.dev20241117+cpu, 0.5.0.dev20241118+cpu, 0.5.0.dev20241119+cpu, 0.5.0.dev20241120+cpu, 0.5.0.dev20241121+cpu, 0.5.0.dev20241126+cpu, 0.5.0.dev20241128+cpu, 0.5.0.dev20241129+cpu, 0.5.0.dev20241130+cpu, 0.5.0.dev20241201+cpu, 0.5.0.dev20241202+cpu, 0.5.0.dev20241203+cpu, 0.5.0.dev20241204+cpu, 0.5.0.dev20241205+cpu, 0.5.0.dev20241206+cpu, 0.5.0.dev20241207+cpu, 0.5.0.dev20241208+cpu, 0.5.0.dev20241209+cpu, 0.5.0.dev20241210+cpu, 0.5.0.dev20241211+cpu, 0.5.0.dev20241212+cpu, 0.5.0.dev20241213+cpu, 0.5.0.dev20241214+cpu, 0.5.0.dev20241215+cpu, 0.5.0.dev20241216+cpu, 0.5.0.dev20241217+cpu, 0.5.0.dev20241218+cpu, 0.5.0.dev20241219+cpu, 0.5.0.dev20241220+cpu, 0.5.0, 0.6.0.dev20241221+cpu, 0.6.0.dev20241222+cpu, 0.6.0.dev20241223+cpu, 0.6.0.dev20241224+cpu, 0.6.0.dev20241225+cpu, 0.6.0.dev20241226+cpu, 0.6.0.dev20241227+cpu, 0.6.0.dev20241228+cpu, 0.6.0.dev20241229+cpu, 0.6.0.dev20241230+cpu, 0.6.0.dev20241231+cpu, 0.6.0.dev20250101+cpu, 0.6.0.dev20250102+cpu, 0.6.0.dev20250103+cpu, 0.6.0.dev20250104+cpu, 0.6.0.dev20250105+cpu, 0.6.0.dev20250106+cpu, 0.6.0.dev20250107+cpu, 0.6.0.dev20250108+cpu, 0.6.0.dev20250109+cpu, 0.6.0.dev20250110+cpu, 0.6.0.dev20250111+cpu, 0.6.0.dev20250112+cpu, 0.6.0.dev20250113+cpu, 0.6.0.dev20250114+cpu, 0.6.0.dev20250115+cpu)
ERROR: No matching distribution found for torchtune==0.4.0.dev20241112
+ cleanup_workspace
+ echo 'sudo may print the following warning message that can be ignored. The chown command will still run.'
sudo may print the following warning message that can be ignored. The chown command will still run.
+ echo '    sudo: setrlimit(RLIMIT_STACK): Operation not permitted'
    sudo: setrlimit(RLIMIT_STACK): Operation not permitted
+ echo 'For more details refer to https://github.com/sudo-project/sudo/issues/42'
For more details refer to https://github.com/sudo-project/sudo/issues/42
+ sudo chown -R 1000 /var/lib/jenkins/workspace
Error: Process completed with exit code 1.
```

as you can see it is related to incorrect specification of a package version `torchtune==0.4.0.dev20241112` in the `examples/models/llama3_2_vision/install_requirements.sh` file of `executorch`.

The bug is seemingly fixed here - https://github.com/pytorch/executorch/commit/a18f6e89631ab97f6211aaae537bfe2cdd90729e#diff-c36c8b70e6987fc4cd000a64a3751bafd2355cf5ceb9a53eea20dcf5d2bc5c54L8
